ftdc/ftdc.go

```diff
@@ -62,9 +121,66 @@ func NewWithWriter(writer io.Writer, logger logging.Logger) *FTDC {
 	}
 }
 
-// newDatum takes an ftdc reading ("Datum") as input and serializes + writes it to the backing
+// Add regsiters a new staters that will be recorded in future FTDC loop iterations.
+func (ftdc *FTDC) Add(name string, statser Statser) {
+	ftdc.mu.Lock()
+	defer ftdc.mu.Unlock()
+
+	for _, statser := range ftdc.statsers {
+		if statser.name == name {
+			ftdc.logger.Warnw("Trying to add conflicting ftdc section", "name", name)
+			// FTDC output is broken down into separate "sections". The `name` is used to label each
+			// section. We return here to predictably include one of the `Add`ed statsers.
+			return
+		}
+	}
+
+	ftdc.logger.Debugw("Added statser", "name", name, "type", fmt.Sprintf("%T", statser))
+	ftdc.statsers = append(ftdc.statsers, namedStatser{
+		name:    name,
+		statser: statser,
+	})
+	ftdc.inputGenerationID++
+}
+
+// Remove removes a statser that was previously `Add`ed with the given `name`.
+func (ftdc *FTDC) Remove(name string) {
+	ftdc.mu.Lock()
+	defer ftdc.mu.Unlock()
+
+	for idx, statser := range ftdc.statsers {
+		if statser.name == name {
+			ftdc.logger.Debugw("Removed statser", "name", name, "type", fmt.Sprintf("%T", statser.statser))
+			ftdc.statsers = append(ftdc.statsers[0:idx], ftdc.statsers[idx+1:len(ftdc.statsers)]...)
```

Comment:
should we use [slices.Delete](https://pkg.go.dev/slices#Delete) here given that (correct me if I'm wrong) if we don't zero out the removed element it won't get GCed?

Reply:
You're right that GCing can be a concern, and I wasn't considering that. (Putting on my pedantic hat) Though it will only leak elements removed at the end. And only if the slice never outgrows its old size. Also I don't think its realistic to think we'll be leaking enough stuff to be a problem. Though we would be leaking random "model" objects which could be expensive.

Practically though, it's much more straightforward to use that `slices` method. I haven't committed to memory that library exists. Thanks for suggesting.